charms/istio-pilot/src/charm.py

```diff
@@ -126,6 +139,58 @@ def remove(self, event):
             manifests, namespace=self.model.name, ignore_not_found=True, ignore_unauthorized=True
         )
 
+    def upgrade_charm(self, event):
+        """Upgrade charm.
+
+        Supports upgrade of exactly one minor version at a time.
+        """
+        istioctl = Istioctl(ISTIOCTL_PATH, self.model.name, ISTIOCTL_DEPOYMENT_PROFILE)
+        self._log_and_set_status(MaintenanceStatus("Upgrading Istio"))
+
+        # Check for version compatability for the upgrade
+        try:
+            versions = istioctl.version()
+        except VersionCheckError as e:
+            self.log.error(UPGRADE_FAILED_MSG.format(message=str(e)))
+            raise GenericCharmRuntimeError("Failed to upgrade.  See `juju debug-log` for details.") from e
+
+        try:
+            _validate_upgrade_version(versions)
+        except ValueError as e:
+            self.log.error(UPGRADE_FAILED_MSG.format(message=str(e)))
+            raise GenericCharmRuntimeError("Failed to upgrade.  See `juju debug-log` for details.") from e
+
+        # TODO: Add an override that can be set by config, eg:
+        #  `juju config istio-pilot skip-upgrade=true`, so we can get out of really sticky
+        #  situations like if our version parsing is wrong54
+
+        # Use istioctl precheck to confirm the upgrade should be safe
+        try:
+            self._log_and_set_status(MaintenanceStatus("Excecuting `istioctl precheck`"))
+            istioctl.precheck()
```

Comment:
Here I would refer to my comments in [wrapper class PR](https://github.com/canonical/istio-operators/pull/191):
`precheck` will return True if existing installation is 1.11 and `istioctl` version we are using is 1.16.2
Not sure if this is right behaviour.